util/include/channels/channel.h

```diff
@@ -0,0 +1,114 @@
+class NoReceiversError : public std::runtime_error {
+ public:
+  NoReceiversError() : std::runtime_error("All receivers have been destructed.") {}
+};
+
+// The interface for the sender side of a channel
+template <typename Msg>
+class NullSender {
+  // Send a message over a channel.
+  //
+  // Sending a message on a channel must be non-blocking to prevent distributed deadlocks as a
+  // result of cyclic topologies. To enable this, if the channel is full we return the message to
+  // the sender. The sender can then determine whether to drop it or buffer it and apply
+  // backpressure.
+  //
+  // Returns std::nullopt when the message is succesfully sent.
+  // Returns Msg when the channel is full.
+  //
+  // Throws NoReceiversError if all receivers have been destructed.
+  //
+  // It's important to note that there is a race condition, whereby a receiver can be destructed
+  // immediately after a successful send. There is no absolute guarantee of delivery. The
+  // purpose of returning an error is to prevent endlessly retrying and never noticing that the
+  // channel is effectively destroyed.
+  std::optional<Msg> send(Msg&& msg) { return std::nullopt; }
+
+  // Return how many elements are in the channels internal buffer. For some channels, only an
+  // estimate will be avaialable, and for some others (mainly lock-free/wait-free) channels, this
+  // value will not be available at all. In those cases std::nullopt should be returned.
+  std::optional<size_t> size() const { return std::nullopt; }
+
+  // Bounded channels should return how many messages they can store in their internal buffer.
+  // Unbounded channels should return std::nullopt;
+  std::optional<size_t> capacity() const { return std::nullopt; }
+};
+
+// The interface for the receiver side of a channel
+template <typename Msg>
+class NullReceiver {
+  // Receive a message over a channel.
+  // Block waiting indefinitely.
+  //
+  // This example implies that Msg must be default constructable. In a real channel this is not
+  // necessary.
+  //
+  // Throws NoSendersError if all senders have been destructed and there are no more messages left
+  // in the communication buffer.
+  Msg recv() { return Msg(); }
+
+  // Receive a message over a channel.
+  // Wait only until the timeout.
+  //
+  // Return std::nullopt if a timeout occurred.
+  std::optional<Msg> recv(std::chrono::milliseconds timeout) { return std::nullopt; };
+
+  // Return how many elements are in the channels internal buffer. For some channels, only an
+  // estimate will be avaialable, and for some others (mainly lock-free/wait-free) channels, this
+  // value will not be available at all. In those cases std::nullopt should be returned.
+  std::optional<size_t> size() { return std::nullopt; }
+
+  // Bounded channels should return how many messages they can store in their internal buffer.
+  // Unbounded channels should return std::nullopt;
+  std::optional<size_t> capacity() { return std::nullopt; }
```

Comment:
Nit: Maybe we can make these `const` so that users can expect it for all channels.